examples/step-58/step-58.cc

```diff
@@ -0,0 +1,673 @@
+  // have updated the `solution` vector $\Psi^n$ to the correct value at the end
+  // of each time step. Its entries contain complex numbers for the solution at
+  // the nodes of the finite element mesh.
+  //
+  // Complex numbers are not easily visualized. We can output their real and
+  // imaginary parts, i.e., the fields $\text{Re}(\psi_h^{(n)}(\mathbf x))$ and
+  // $\text{Im}(\psi_h^{(n)}(\mathbf x))$, and that is exactly what the DataOut
+  // class does when one attaches as complex-valued vector via
+  // DataOut::add_data_vector() and then calls DataOut::build_patches(). That is
+  // indeed what we do below.
+
+  // But oftentimes we are not particularly interested in real and imaginary
+  // parts of the solution vector, but instead in derived quantities such as the
+  // magnitude $|\psi|$ and phase angle $\text{arg}(\psi)$ of the solution. In
+  // the context of quantum systems such as here, the magnitude itself is not so
+  // interesting, but instead it is the "amplitude", $|\psi|^2$ that is a
+  // physical property: it corresponds to the probability density of finding a
+  // particle in a particular place of state. The way to put computed quantities
+  // into output files for visualization -- as used in numerous previous
+  // tutorial programs -- is to use the facilities of the DataPostprocessor and
+  // derived classes. Specifically, both the amplitude of a complex number and
+  // its phase angles are scalar quantities, and so the DataPostprocessorScalar
+  // class is the right tool to base what we want to do on.
+  //
+  // Consequently, what we do here is to implement two classes
+  // `ComplexAmplitude` and `ComplexPhase` that compute for each point at which
+  // DataOut decides to generate output, the amplitudes $|\psi_h|^2$ and phases
+  // $\text{arg}(\psi_h)$ of the solution for visualization. There is a fair
+  // amount of boiler-plate code below, with the only interesting parts of
+  // the first of these two classes being how its `evaluate_vector_field()`
+  // function computes the `computed_quantities` object.
+  //
+  // (There is also the rather awkward fact that the <a
+  // href="https://en.cppreference.com/w/cpp/numeric/complex/norm">std::norm()</a>
+  // function does not compute what one would naively imagine, namely $|\psi|$,
+  // but returns $|\psi|^2$ instead. It's certainly quite confusing to have a
+  // standard function mis-named in such a way...)
+  namespace DataPostprocessors
+  {
+    template <int dim>
+    class ComplexAmplitude : public DataPostprocessorScalar<dim>
+    {
+    public:
+      ComplexAmplitude();
+
+      virtual void evaluate_vector_field(
+        const DataPostprocessorInputs::Vector<dim> &inputs,
+        std::vector<Vector<double>> &computed_quantities) const override;
+    };
+
+
+    template <int dim>
+    ComplexAmplitude<dim>::ComplexAmplitude()
+      : DataPostprocessorScalar<dim>("Magnitude", update_values)
```

Comment:
Out of curiosity, is there a reason as to why the class is named `ComplexAmplitude` but the title that you embed in the output is `Magnitude`? I mean, it doesn't really matter and both seem fine, but the inconsistency is a bit unexpected.

Reply:
Nope, it's a mistake. I used to output the *magnitude* and forgot to change this one occurrence.